src/Exceptionless.Core/Services/StackService.cs

```diff
@@ -0,0 +1,94 @@
+
+            var cachedOccurrenceMinDateUtc = await _cache.GetAsync<DateTime>(occurrenceMinDateCacheKey).AnyContext();
+            if (!cachedOccurrenceMinDateUtc.HasValue || cachedOccurrenceMinDateUtc.IsNull || cachedOccurrenceMinDateUtc.Value > minOccurrenceDateUtc)
+                tasks.Add(_cache.SetAsync(occurrenceMinDateCacheKey, minOccurrenceDateUtc, expireTimeout));
+
+            var cachedOccurrenceMaxDateUtc = await _cache.GetAsync<DateTime>(occurrenceMaxDateCacheKey).AnyContext();
+            if (!cachedOccurrenceMaxDateUtc.HasValue || cachedOccurrenceMinDateUtc.IsNull || cachedOccurrenceMaxDateUtc.Value < maxOccurrenceDateUtc)
+                tasks.Add(_cache.SetAsync(occurrenceMaxDateCacheKey, maxOccurrenceDateUtc, expireTimeout));
+
+            tasks.Add(_cache.IncrementAsync(occurenceCountCacheKey, count, expireTimeout));
+            tasks.Add(_cache.SetAddAsync(occurrenceSetCacheKey, Tuple.Create(organizationId, projectId, stackId), expireTimeout));
+
+            await Task.WhenAll(tasks).AnyContext();
+        }
+
+        public async Task SaveStackUsagesAsync(bool sendNotifications = true, CancellationToken cancellationToken = default) {
+            var occurrenceSetCacheKey = GetStackOccurrenceSetCacheKey();
+            var stackUsageSet = await _cache.GetSetAsync<Tuple<string, string, string>>(occurrenceSetCacheKey).AnyContext();
+            if (!stackUsageSet.HasValue || stackUsageSet.IsNull) return;
+            foreach (var tuple in stackUsageSet.Value) {
+                if (cancellationToken.IsCancellationRequested) break;
+
+                string organizationId = tuple.Item1, projectId = tuple.Item2, stackId = tuple.Item3;
+                string occurenceCountCacheKey = GetStackOccurrenceCountCacheKey(organizationId, projectId, stackId),
+                    occurrenceMinDateCacheKey = GetStackOccurrenceMinDateCacheKey(organizationId, projectId, stackId),
+                    occurrenceMaxDateCacheKey = GetStackOccurrenceMaxDateCacheKey(organizationId, projectId, stackId);
+                var occurrenceCount = await _cache.GetAsync<long>(occurenceCountCacheKey, 0).AnyContext();
+                if (occurrenceCount == 0) return;
+                var occurrenceMinDate = _cache.GetAsync(occurrenceMinDateCacheKey, SystemClock.UtcNow);
+                var occurrenceMaxDate = _cache.GetAsync(occurrenceMaxDateCacheKey, SystemClock.UtcNow);
+
+                await Task.WhenAll(occurrenceMinDate, occurrenceMaxDate).AnyContext();
+
+                var tasks = new List<Task> {
+                    _stackRepository.IncrementEventCounterAsync(organizationId, projectId, stackId, occurrenceMinDate.Result, occurrenceMaxDate.Result, (int)occurrenceCount, sendNotifications),
+                    _cache.RemoveAllAsync(new[] { occurenceCountCacheKey, occurrenceMinDateCacheKey, occurrenceMaxDateCacheKey }),
+                    _cache.SetRemoveAsync(occurrenceSetCacheKey, tuple, TimeSpan.FromHours(24))
```

Comment:
We gotta think about this as I know we can't do it this way due to race conditions. We could be removing values even from the time we were getting values from the cache (that were being added after we got our value). We gotta think of a way to do this without locking on incrementing as that would slow down the pipeline lots too. 

Not sure what the FromHours(24) is here. I wonder if we could just pop an item from a list and get the value and then remove that, that way we know anything in that list was old? @ejsmith @edwardmeng thoughts?